plugins/in_dbus/in_dbus.c

```diff
@@ -0,0 +1,527 @@
+/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*  Fluent Bit
+ *  ==========
+ *  Copyright (C) 2019      The Fluent Bit Authors
+ *  Copyright (C) 2015-2018 Treasure Data Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <fluent-bit/flb_info.h>
+#include <fluent-bit/flb_input.h>
+#include <fluent-bit/flb_config.h>
+#include <fluent-bit/flb_error.h>
+#include <fluent-bit/flb_utils.h>
+#include <fluent-bit/flb_pack.h>
+#include <fluent-bit/flb_parser.h>
+#include <msgpack.h>
+#include <dbus/dbus.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "in_dbus.h"
+
+struct in_dbus_event {
+    struct mk_event event;
+    struct mk_event_loop* evl;
+    DBusConnection* conn;
+    DBusWatch* watch;
+};
+
+int in_dbus_event_handler(void* data) {
+    struct in_dbus_event* event = data;
+    const unsigned flags = dbus_watch_get_flags(event->watch);
+    if (!dbus_watch_handle(event->watch, flags)) {
+        flb_error("[in_dbus] dbus_watch_handle failed");
+    }
+
+    while (dbus_connection_get_dispatch_status(event->conn) ==
```

Comment:
Since the dbus dispatching now longer happens in a dedicated thread, we may want to consider bounding this loop to avoid tying up the main event loop processing for too long.